### PR TITLE
SoundSourceFFmpeg: Fix debug assert and offset/range calculation

### PR DESCRIPTION
**Pull request by uklotzde**
- 1st commit is only a reformatting
- Debug assertion: noticed while testing #2911
- The range/offset calculations affects files encoded with iTunes 12.3.0 that include a `start_time` were wrong. Comparison with the FFmpeg CLI revealed the differences.
- Files encoded with iTunes 12.7.0 always have this ~48 ms (2112 frames at 44.1 kHz) offset and longer duration, independent of the decoder (`start_time` is missing)

I also noticed that FAAD2 handles the affected AAC files differently and generates both an offset with silence at the start and the duration is longer than expected. But I won't mirror the wrong behavior of FAAD2.

Next step: Test the different MP3 encodings and compare with FFmpeg CLI/libMAD/MPG123.

### CODE REVIEW

**Comment by uklotzde** [comment]
More code cleanup will follow after this PR has been merged.

File: src/sources/soundsourceffmpeg.cpp [hunk 1]
```diff
@@ -16,7 +16,7 @@ constexpr AVSampleFormat kavSampleFormat = AV_SAMPLE_FMT_FLT;
 
 constexpr uint64_t kavChannelLayoutUndefined = 0;
 
-constexpr int64_t kavMinStartTime = 0;
+constexpr int64_t kavDefaultStreamStartTime = 0;
```
Comment (Holzhaus) [inline src/sources/soundsourceffmpeg.cpp:19]: What is the unit? Frame Indices?

Comment (Be-ing) [reply src/sources/soundsourceffmpeg.cpp:19]: I am looking forward to having the FramePos class from #2861 to use in the rest of Mixxx.

Comment (uklotzde) [reply src/sources/soundsourceffmpeg.cpp:19]: No, FFmpeg time units. That's why it is prefixed with "av".

Comment (Holzhaus) [reply src/sources/soundsourceffmpeg.cpp:19]: Until we have that we should add the unit it to the variable name. That makes it easier to spot issues during code review.

Comment (uklotzde) [reply src/sources/soundsourceffmpeg.cpp:19]: It is named in the same way as the corresponding functions with the suffix "Time". The type also mirrors the FFmpeg type. I don't know of a better name.

Comment (Holzhaus) [reply src/sources/soundsourceffmpeg.cpp]: @uklotzde sorry I didn't see your comment when I replied. In that case it can stay like that.

**Comment by uklotzde** [comment]
AppVeyor build timed out, unrelated.

**Comment by Holzhaus** [comment]
I tried to test his by dragging an M4A file from the builtin file browser to a deck, but then Mixxx aborted with:

```
DEBUG ASSERT: "pCue->getTrackId() == trackId" in function void CueDAO::saveTrackCues(TrackId, const QList<CuePointer>&) const at /home/jan/Projects/mixxx/src/library/dao/cuedao.cpp:220
```

Probably unrelated to this branch though...

**Comment by uklotzde** [comment]
> I tried to test his by dragging an M4A file from the builtin file browser to a deck, but then Mixxx aborted with:
> 
> ```
> DEBUG ASSERT: "pCue->getTrackId() == trackId" in function void CueDAO::saveTrackCues(TrackId, const QList<CuePointer>&) const at /home/jan/Projects/mixxx/src/library/dao/cuedao.cpp:220
> ```
> 
> Probably unrelated to this branch though...

Unrelated.

We need to find out when this debug assertion is triggered. Now we know that this happens when dragging files from the browser! Maybe the addition to the library happens too late and after trying to store the cue points?

**Comment by uklotzde** [comment]
Idea: Add an assertion when adding cue points to tracks. This should only be allowed when the `Track` object has an id. Otherwise inconsistencies occur.

One more reason to not handle `Cue` as separate QObjects. They should be simple, passive C++ objects that are managed by the `Track` object.

**Comment by Holzhaus** [comment]
> Idea: Add an assertion when adding cue points to tracks. This should only be allowed when the `Track` object has an id. Otherwise inconsistencies occur.
> 
> One more reason to not handle `Cue` as separate QObjects. They should be simple, passive C++ objects that are managed by the `Track` object.

Added a debug assertion that triggers when I open a directory in the file browser that contains tracks with Serato Cue points in their metadata:
```
DEBUG ASSERT: "getId().isValid()" in function void Track::setCuePointsMarkDirtyAndUnlock(QMutexLocker*, const QList<CuePointer>&) at /home/jan/Projects/mixxx/src/track/track.cpp:950
```

**Comment by uklotzde** [comment]
> > Idea: Add an assertion when adding cue points to tracks. This should only be allowed when the `Track` object has an id. Otherwise inconsistencies occur.
> > One more reason to not handle `Cue` as separate QObjects. They should be simple, passive C++ objects that are managed by the `Track` object.
> 
> Added a debug assertion that triggers when I open a directory in the file browser that contains tracks with Serato Cue points in their metadata:
> 
> ```
> DEBUG ASSERT: "getId().isValid()" in function void Track::setCuePointsMarkDirtyAndUnlock(QMutexLocker*, const QList<CuePointer>&) at /home/jan/Projects/mixxx/src/track/track.cpp:950
> ```

Maybe not need if this fix works: #2934

**Comment by uklotzde** [comment]
Ping

**Comment by uklotzde** [comment]
It is essential to get the decoding and offsets right now that Ubuntu 20.04 actually uses SoundSourceFFmpeg!

**Review by Holzhaus — COMMENTED** [review]
Works fine and tests are fixed, but I cannot really comment on the offset calculation as I don't have MP4 files in my library.

**Comment by uklotzde** [comment]
Ping again.

The longer these changes are hold back the more users on Ubuntu 20.04 will experience offsets with .m4a files that include a start_time. I noticed various in my library where cue positions need to be fixed.

**Comment by uklotzde** [comment]
Last ping. Otherwise I will close the PR and only maintain a private fork for myself in the future. This is frustrating.

**Comment by Holzhaus** [comment]
This branch works fine for me and I trust that you tested this thoroughly. Let's give it a few more days, and if nobody complains then I'll merge this.

**Comment by Holzhaus** [comment]
Is someone interested in reviewing this? Otherwise I'll merge this tomorrow.

**Comment by Be-ing** [comment]
I am unfamiliar with this code and don't understand it well, so I do not intend to review it.

**Pull request merged**